sorting/count_inversions.cpp

```diff
@@ -0,0 +1,268 @@
+/**
+ * @file
+ * @brief Counting Inversions using [Merge
+ Sort](https://en.wikipedia.org/wiki/Merge_sort)
+ *
+ * @details
+ * Program to count the number of inversions in an array
+ * using merge-sort.
+ *
+ * The count of inversions help to determine how close the array
+ * is to being sorted in ASCENDING order.
+ *
+ * two elements a[i] and a[j] form an inversion if a[i] > a[j] and i < j
+ *
+ * Time Complexity --> O(n.log n)
+
+ * Space Complexity --> O(n) ; additional array temp[1..n]
+ * ### Algorithm
+
+ *   1. The idea is similar to merge sort, divide the array into two equal or
+ almost
+ *      equal halves in each step until the base case is reached.
+ *   2. Create a function merge that counts the number of inversions when two
+ halves of
+ *      the array are merged, create two indices i and j, i is the index for
+ first half
+ *      and j is an index of the second half. if a[i] is greater than a[j], then
+ there are (mid – i)
+ *      inversions, Because left and right subarrays are sorted, so all the
+ remaining elements
+ *      in left-subarray (a[i+1], a[i+2] … a[mid]) will be greater than a[j].
+ *   3. Create a recursive function to divide the array into halves and find the
+ answer by summing
+ *      the number of inversions is the first half, number of inversion in the
+ second half and
+ *      the number of inversions by merging the two.
+ *   4. The base case of recursion is when there is only one element in the
+ given half.
+ *   5. Print the answer
+ *
+ * @author [Rakshit Raj](https://github.com/rakshitraj)
+ */
+#include <cassert>
+#include <iostream>
+#include <vector>
+
+/**
+ * @namespace sorting
+ * @brief Sorting algorithms
+ */
+namespace sorting {
+/**
+ * @namespace inversion
+ * @brief Functions for counting inversions using Merge Sort algorithm
+ */
+namespace inversion {
+
+// Functions used --->
+// int mergeSort(int* arr, int* temp, int left, int right);
+// int merge(int* arr, int* temp, int left, int mid, int right);
+// int countInversion(int* arr, const int size);
+// void show(int* arr, const int size);
+
+/**
+ * Function to merge two sub-arrays. merge() function is called
+ * from mergeSort() to merge the array after it split for sorting
+ * by the mergeSort() funtion.
+ *
+ * In this case the merge fuction will also count and return
+ * inversions detected when merging the sub arrays.
+ *
+ * @param arr    input array, data-menber of vector
+ * @param temp   stores the resultant merged array
+ * @param left   lower bound of arr[] and left-sub-array
+ * @param mid    midpoint, upper bound of left sub-array,
+ *               (mid+1) gives the lower bound of right-sub-array
+ * @param right  upper bound of arr[] and right-sub-array
+ * @returns number of inversions found in merge step
+ */
+template<typename T>
+int merge(T* arr, T* temp, int left, int mid, int right) {
```

Comment:
It is recommended to use `uint64_t` for non-negative values or `int64_t` for negative values (or their appropriate size). [Reference](https://en.cppreference.com/w/cpp/types/integer).